hw7/searchTree/tree.c

```diff
@@ -0,0 +1,231 @@
+            if (node->leftSon == NULL)
+            {
+                *isWorking = false;
+                free(strValue);
+                return;
+            }
+            node->leftSon->key = key;
+            node->leftSon->value = strValue;
+            return;
+        }
+        addValueRecursive(node->leftSon, key, value, isWorking);
+    }
+    if (node->key < key)
+    {
+        if (node->rightSon == NULL)
+        {
+            node->rightSon = calloc(sizeof(Node), 1);
+            if (node->rightSon == NULL)
+            {
+                *isWorking = false;
+                free(strValue);
+                return;
+            }
+            node->rightSon->key = key;
+            node->rightSon->value = strValue;
+            return;
+        }
+        addValueRecursive(node->rightSon, key, strValue, isWorking);
+    }
+}
+
+void addValue(Tree* tree, int key, const char* value, bool* isWorking)
+{
+    if (!isEmpty(tree))
+    {
+        addValueRecursive(tree->root, key, value, isWorking);
+        return;
+    }
+    char* strValue = calloc(strlen(value) + 1, sizeof(char));
+    if (strValue == NULL)
+    {
+        *isWorking = false;
+        free(strValue);
+        return;
+    }
+    strcpy(strValue, value);
+    tree->root = calloc(sizeof(Node), 1);
+    if (tree->root == NULL)
+    {
+        *isWorking = false;
+        return;
+    }
+    tree->root->key = key;
+    tree->root->value = strValue;
+}
+
+bool inTree(Tree* tree, int key)
+{
+    Node* i = tree->root;
+    while (i != NULL)
+    {
+        if (key > i->key)
+        {
+            i = i->rightSon;
+        }
+        else if (key < i->key)
+        {
+            i = i->leftSon;
+        }
+        else
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+char* getValueRecursive(Node* node, int key)
+{
+    if (node == NULL)
+    {
+        return NULL;
+    }
+    if (node->key == key)
+    {
+        return node->value;
+    }
+    if (node->key < key)
+    {
+        return getValueRecursive(node->rightSon, key);
+    }
+    return getValueRecursive(node->leftSon, key);
+}
+
+char* getValue(Tree* tree, int key)
+{
+    return getValueRecursive(tree->root, key);
+}
+
+Node* subsequentNode(Node* node)
+{
+    Node* right = node->rightSon;
+    while (right->leftSon != NULL)
+    {
+        right = right->leftSon;
+    }
+    return right;
+}
+
+Node* deleteNode(Node* node, int key, bool* isWorking)
+{
+    if (node == NULL)
+    {
+        return NULL;
+    }
+    if (node->key == key)
+    {
+        if (node->leftSon == NULL && node->rightSon == NULL)
+        {
+            free(node->value);
+            free(node);
+            return NULL;
+        }
+        else if (node->leftSon == NULL || node->rightSon == NULL)
+        {
+            Node* child = (node->leftSon != NULL ? node->leftSon : node->rightSon);
+            free(node->value);
+            free(node);
+            node = child;
+            return node;
```

Comment:
```suggestion
            return child;
```
Изменение значения параметра node всё равно никто никогда не увидит, можете хоть `node = (Node*)666;` написать :)